qemu/tests/numa_opts.py

```diff
@@ -14,36 +14,99 @@ def run(test, params, env):
+    def numa_info_guest():
+        """
+        Run 'numactl -H' in guest os and parse returned information, linux only
+
+        return: An array of (ram, cpus) tuples, where ram is the RAM size in
+                MB and cpus is a set of CPU numbers
+        """
+        session = vm.wait_for_login()
+        numa_guest_info = session.cmd_output('numactl -H')
+        re_numa_nodes = re.compile(r"([0-9]+) nodes", re.M)
+        re_numa_node_info = re.compile(r"^node ([0-9]+) (cpus|size): (.*)$", re.M)
+        nodes = re_numa_nodes.search(numa_guest_info)
+        # At least one node in vm, even no extra parameters
+        if nodes is None:
+            test.error("Couldn't get number of nodes from 'numactl -H' output")
+        nodes = int(nodes.group(1))
+        numa_guest = [[0, set()] for i in range(nodes)]
+        for nodenr, field, value in re_numa_node_info.findall(numa_guest_info):
+            nodenr = int(nodenr)
+            if nodenr > nodes:
+                test.fail("Invalid node num on 'numactl -H' output: %d", nodenr)
+            if field == 'size':
+                if not value.endswith(' MB'):
+                    test.error("Unexpected size value: %s", value)
+                megabytes = int(value[:-3])
+                numa_guest[nodenr][0] = megabytes
+            elif field == 'cpus':
+                cpus = set([int(v) for v in value.split()])
+                numa_guest[nodenr][1] = cpus
+        numa_guest = [tuple(i) for i in numa_guest]
+        # It is a known WONTFIX issue, node info of node0 and node1 is opposite
+        # with 'numactl -H' when vm have 2 nodes
+        if len(numa_guest) == 2:
+            numa_guest.reverse()
+        return numa_guest
 
     vm = env.get_vm(params["main_vm"])
+    os_type = params["os_type"]
+
+    error_context.context("starting numa_opts test...", logging.info)
 
+    # Get numa info from monitor with "info monitor"
     numa = vm.monitors[0].info_numa()
-    dbg("info numa reply: %r", numa)
+    error_context.context("info numa reply: %r" % numa, logging.info)
+
+    if os_type == 'linux':
+        # Get numa info in vm with "numactl -H",  only for Linux
+        numa_guest = numa_info_guest()
+        error_context.context("numactl -H reply: %r" % numa_guest, logging.info)
+    else:
+        numa_guest = numa
 
     numa_nodes = params.get("numa_nodes")
     if numa_nodes:
-        numa_nodes = int(params.get("numa_nodes"))
+        numa_nodes = int(numa_nodes)
         if len(numa) != numa_nodes:
-            test.fail(
-                "Wrong number of numa nodes: %d. Expected: %d" %
-                (len(numa), numa_nodes))
+            test.fail("[Monitor]Wrong number of numa nodes: %d. Expected: %d" %
+                      (len(numa), numa_nodes))
+
+    if os_type == 'linux':
+        g_numa_nodes = params.get("g_numa_nodes", numa_nodes)  # min value is 1
+        g_numa_nodes = int(g_numa_nodes)
+        if len(numa_guest) != g_numa_nodes:
```

Comment:
How about not introducing new param? The if condition is like this:
if len(numa_guest) ! = numa_nodes>0 or (len(numa_guest)!=1 and numa_nodes==0)

Reply:
Yes,  without the new param we can achieve the same result. However, isn't is a little convoluted for user to get the expect result  that `no extra parameters => one node with 'numactl -H' in guest` ?

Reply:
I think it's same to user whether it is in cfg or py. Btw, the param name is not very informative.

Reply:
> I think it's same to user whether it is in cfg or py. Btw, the param name is not very informative.

Could you give is a more informative name ? I prefer to keep the new param ~ Complex comparison statement is not easy to understand for new guys to get the expect result.

Reply:
How about 'guest_expected_nodes'?

Reply:
> How about 'guest_expected_nodes'?

Make sense, updated~  thanks!